### PR TITLE
Mobile: Resolves #468: Add the ability to record audio and attach it to a note

### LINKED ISSUE #468: [Feature Request] Audio notes
Recording voice notes and storing them in the cloud.

### PR DESCRIPTION
**Pull request by bryangerlach**
I added the ability to record audio from within a note and attach it.  From within a note, open the menu and select Record Audio (or click Attach... and select Record Audio, we can keep whichever one is best):

Screenshot_1586639007
OR:
device-2020-04-11-161024

Then when you are done recording click the stop button (square button):

device-2020-04-11-160838

Then the file is attached to the note:

device-2020-04-11-160902
device-2020-04-11-160918

### CODE REVIEW

**Comment by laurent22** [comment]
Thanks for the pull request. Is it for Android only? If so it will need to be tested on iOS anyway to make sure it doesn't break anything as I see you import several Android in a screen that's used by the iOS app too.

File: CliClient/locales/en_US.po [hunk 1]
```diff
@@ -3214,6 +3214,10 @@ msgstr ""
 msgid "Take photo"
 msgstr ""
 
+#: /Users/tessus/data/work/joplin/Tools/../ReactNativeClient/lib/components/screens/note.js:710
```
Comment (laurent22) [inline CliClient/locales/en_US.po:3217]: Please undo the changes in this file

File: ReactNativeClient/lib/components/screens/note.js [hunk 2]
```diff
@@ -604,6 +625,101 @@ class NoteScreenComponent extends BaseScreenComponent {
...
+				uri: result,
+				didCancel: false,
+				error: null,
+				type: 'audio/mp4',
+				fileName: 'joplin-'+new Date().getMonth()+'.'+new Date().getDate()+'.'+new Date().getFullYear()+' - '+new Date().getHours()+'.'+new Date().getMinutes(),
+			},
+			'audio'
+		);
+		this.setState({ isrecording: false });
+	};
```
Comment (laurent22) [inline ReactNativeClient/lib/components/screens/note.js:721]: Could you wrap all the audio related logic in a service in lib/service?

File: ReactNativeClient/lib/components/screens/note.js [hunk 3]
```diff
@@ -604,6 +625,101 @@ class NoteScreenComponent extends BaseScreenComponent {
...
+					},
+			  	);
+				if (granted === PermissionsAndroid.RESULTS.GRANTED) {
+					console.log('You can use the storage');
+				} else {
+					console.log('permission denied');
+					return;
+				}
+			} catch (err) {
+				console.warn(err);
```
Comment (laurent22) [inline ReactNativeClient/lib/components/screens/note.js:658]: If something doesn't work we need to display errors to the user.

File: ReactNativeClient/lib/components/screens/note.js [hunk 4]
```diff
@@ -604,6 +625,101 @@ class NoteScreenComponent extends BaseScreenComponent {
...
+				if (granted === PermissionsAndroid.RESULTS.GRANTED) {
+					console.log('You can use the camera');
+				} else {
+					console.log('permission denied');
+					return;
+				}
+			} catch (err) {
+				console.warn(err);
+				return;
+			}
```
Comment (laurent22) [inline ReactNativeClient/lib/components/screens/note.js:681]: The above seems to repeat the logic just above it or am I missing something?

File: ReactNativeClient/package.json [hunk 5]
```diff
@@ -50,10 +50,11 @@
     "prop-types": "^15.6.0",
     "punycode": "^2.1.1",
     "query-string": "4.3.4",
-    "react": "16.9.0",
+    "react": "16.11.0",
     "react-async": "^10.0.0",
-    "react-native": "0.61.5",
+    "react-native": "0.62.2",
```
Comment (laurent22) [inline ReactNativeClient/package.json:55]: Please remove

File: ReactNativeClient/package.json [hunk 6]
```diff
@@ -50,10 +50,11 @@
     "prop-types": "^15.6.0",
     "punycode": "^2.1.1",
     "query-string": "4.3.4",
-    "react": "16.9.0",
+    "react": "16.11.0",
```
Comment (laurent22) [inline ReactNativeClient/package.json:53]: Please remove

**Review by laurent22 — CHANGES_REQUESTED** [review]

File: ReactNativeClient/lib/components/screens/note.js (hunk 4 above)
Comment (bryangerlach) [reply ReactNativeClient/lib/components/screens/note.js:681]: It is checking for 2 different permissions. It needs WRITE_EXTERNAL_STORAGE and RECORD_AUDIO. I need to update the messages though.

File: ReactNativeClient/lib/components/screens/note.js (hunk 2 above)
Comment (bryangerlach) [reply ReactNativeClient/lib/components/screens/note.js]: I don't exactly know what I'm doing, but I attempted what you have asked here.

File: ReactNativeClient/lib/services/RecordAudio.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	static instance() {
+		if (RecordAudio.instance_) return RecordAudio.instance_;
+		RecordAudio.instance_ = new RecordAudio();
+		return RecordAudio.instance_;
+	}
+
+	async onStartRecord() {
+		if (Platform.OS === 'android') {
+			try {
+				const granted = await PermissionsAndroid.request(
```
Comment (roman-r-m) [inline ReactNativeClient/lib/services/RecordAudio.js:21]: PermissionsAndroid might not work correctly: https://github.com/laurent22/joplin/blob/142976f012d9398bfd2e715bd6fb4bf32dc7b5d5/ReactNativeClient/lib/components/screens/config.js#L468

And indeed on my phone it always returns `never_ask_again`

react-native-permissions is a good alternative.

File: ReactNativeClient/lib/services/RecordAudio.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
+import AudioRecorderPlayer from 'react-native-audio-recorder-player';
```
Comment (laurent22) [inline ReactNativeClient/lib/services/RecordAudio.js:1]: Please check this file and remove the code you're not using. I've noticed dispatch, and the syncCalls and destroy method, which were probably copied and pasted from somewhere else. There might be more so please check.

Also since this is a new file, please use TypeScript as mentioned there: https://github.com/laurent22/joplin/blob/master/CONTRIBUTING.md#coding-style

Comment (laurent22) [reply ReactNativeClient/lib/services/RecordAudio.js:1]: Finally, please move this file to `ReactNativeClient/services` as it's for mobile only (while what's in "lib" is generally shared)

**Review by laurent22 — CHANGES_REQUESTED** [review]

File: ReactNativeClient/lib/services/RecordAudio.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+					console.log('You can record audio');
+				} else {
+					console.log('permission denied');
+					return;
+				}
+			} catch (err) {
+				console.warn(err);
+				Alert.alert('Permission Needed', 'Unable to get microphone permission.');
+				return;
+			}
```
Comment (laurent22) [inline ReactNativeClient/lib/services/RecordAudio.js:61]: I'm not so keen on big nested if/else statements like this one. Could you perhaps refactor into a function as is done in screens/config.js. There's also the issue mentioned by @roman-r-m that PermissionAndroid doesn't work well but we can leave it for now. If it's separated in a separate function it will be easier to change the permission package later on.

Comment (roman-r-m) [reply ReactNativeClient/lib/services/RecordAudio.js:61]: Good point, thanks. I actually wanted to extract this since I used the same code in another PR here: https://github.com/laurent22/joplin/pull/3202

File: ReactNativeClient/lib/services/RecordAudio.js (hunk 7 above)
Comment (roman-r-m) [reply ReactNativeClient/lib/services/RecordAudio.js]: Ignore that, tested again and PermissionsAndroid works now.

**Comment by sunjam** [comment]
Is audio recorded to a particular encoding?  wav, ogg, mp3, [opus](https://opus-codec.org/), etc.

**Comment by laurent22** [comment]
Hello, I'm closing this pull request as I think specialised features like this one should eventually be implemented as plugins (once the plugin system is ready). Of course this pull request could be used as a starting point to implement such plugin in the future.

**Comment by battaglia01** [comment]
Hi @laurent22 - *can* this be implemented in the plugin system? This is a pull request for *mobile*, and the plugin system is for desktop only, right? I am maybe interested in doing something with this, but I'm not sure where to begin.

**Comment by nomatica** [comment]
What is the state of this feature? Is it possible on desktop and android apps?

**Comment by CntrlX** [comment]
@laurent22 what is the status of the plugin system?
